From 1e8e33fb467a9df375d1fd4b250ab40c8d340be1 Mon Sep 17 00:00:00 2001
From: Chisoon Jeong <chisoon.jeong@lge.com>
Date: Mon, 22 Apr 2024 13:49:22 +0900
Subject: [PATCH] Support a doubly linked timezone file path used for webOS

:Release Notes:
Supporting a doubly linked timezone file path used for webOS

:Detailed Notes:
Support both doubly linked timezone path used for webOS and
macOS' timezone file path using /zoneinfo.default/XXX/XXX
instead of /zoneinfo/XXX/XXX that is commonly used for Unix
like system.

Releavent ICU upstream changes:
ICU-21386 uprv_tzname() should find the correct Olson ID
when /etc/localtime is a "double" link
- https://github.com/unicode-org/icu/pull/2323/files
ICU-22541 Fix MacOS 14 default timezone issue #2669
- https://github.com/unicode-org/icu/pull/2669/files

:Test Scenario:
1. modifiy webapp-mgr.sh to put --enable-logging=stderr --v=1
   to commandline flag of WebAppMgr

2. append "2>&1 &> /tmp/wam.log" to exec command to execute WebAppMgr
exec $WAM_EXE_PATH $WAM_SWITCHES $WAM_EXTRA_SKIA_CACHE_SWITCHES \
     $WAM_EXTRA_GPU_TUNING_SWITCHES \
     $WAM_WATCHDOG_RENDERER_TIMEOUT_SWITCHES $WAM_MEM_FLAGS \
     $WAM_EXTRA_FLAGS --webos-js-flags="$WAM_JS_FLAGS" \
     & 2>&1 &> /tmp/wam.log

3. target $ restart webapp-mgr ; tail -F /tmp/wam.log | grep timezone

4. Launch Settings app. and change the Time Zone to
   Europe/Dublin

5. Verify that the time is displayed correctly on
   Settings app.

6. /tmp/wam.log print the correct timezone that is not
  an abbreviation.
Expected result
timezone reset to Europe/Dublin
Issue
timezone reset to IST

:QA Notes:

:Issues Addressed:
[NEVA-7547] Provide the correct location information for Neva Appruntime
---
 source/common/putil.cpp    | 33 ++++++++++++++++++++++++---------
 source/common/uposixdefs.h |  2 +-
 2 files changed, 25 insertions(+), 10 deletions(-)

diff --git a/source/common/putil.cpp b/source/common/putil.cpp
index 83f5cf49..25376d73 100644
--- a/source/common/putil.cpp
+++ b/source/common/putil.cpp
@@ -1171,16 +1171,31 @@ uprv_tzname(int n)
         because the tzfile contents is underspecified.
         This isn't guaranteed to work because it may not be a symlink.
         */
-        int32_t ret = (int32_t)readlink(TZDEFAULT, gTimeZoneBuffer, sizeof(gTimeZoneBuffer)-1);
-        if (0 < ret) {
+        char *ret = realpath(TZDEFAULT, gTimeZoneBuffer);
+        if (ret != nullptr && uprv_strcmp(TZDEFAULT, gTimeZoneBuffer) != 0) {
             int32_t tzZoneInfoTailLen = uprv_strlen(TZZONEINFOTAIL);
-            gTimeZoneBuffer[ret] = 0;
-            char *  tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
-
-            if (tzZoneInfoTailPtr != NULL
-                && isValidOlsonID(tzZoneInfoTailPtr + tzZoneInfoTailLen))
-            {
-                return (gTimeZoneBufferPtr = tzZoneInfoTailPtr + tzZoneInfoTailLen);
+            char*  tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
+            // MacOS14 has the realpath as something like
+            // /usr/share/zoneinfo.default/Australia/Melbourne
+            // which will not have "/zoneinfo/" in the path.
+            // Therefore if we fail, we fall back to read the link which is
+            // /var/db/timezone/zoneinfo/Australia/Melbourne
+            // We also fall back to reading the link if the realpath leads to something like
+            // /usr/share/zoneinfo/posixrules
+            if (tzZoneInfoTailPtr == nullptr ||
+                    uprv_strcmp(tzZoneInfoTailPtr + tzZoneInfoTailLen, "posixrules") == 0) {
+                ssize_t size = readlink(TZDEFAULT, gTimeZoneBuffer, sizeof(gTimeZoneBuffer)-1);
+                if (size > 0) {
+                    gTimeZoneBuffer[size] = 0;
+                    tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
+                }
+            }
+            if (tzZoneInfoTailPtr != nullptr) {
+                tzZoneInfoTailPtr += tzZoneInfoTailLen;
+                skipZoneIDPrefix(const_cast<const char**>(&tzZoneInfoTailPtr));
+                if (isValidOlsonID(tzZoneInfoTailPtr)) {
+                    return (gTimeZoneBufferPtr = tzZoneInfoTailPtr);
+                }
             }
         } else {
 #if defined(SEARCH_TZFILE)
diff --git a/source/common/uposixdefs.h b/source/common/uposixdefs.h
index 3770a335..7c008fdd 100644
--- a/source/common/uposixdefs.h
+++ b/source/common/uposixdefs.h
@@ -48,7 +48,7 @@
 #endif
 
 /*
- * Make sure things like readlink and such functions work.
+ * Make sure things like realpath and such functions work.
  * Poorly upgraded Solaris machines can't have this defined.
  * Cleanly installed Solaris can use this #define.
  *
-- 
2.25.1
